File: src/realm/RealmServer.java
package realm;

import java.io.IOException;
import java.net.ServerSocket;
import java.util.ArrayList;

import common.Main;

public class RealmServer implements Runnable {

	private ServerSocket _SS;
	private Thread _t;
	private final ArrayList<RealmThread> _clients = new ArrayList<>();

	public RealmServer()
	{
		try
		{
			_SS = new ServerSocket(Main.REALM_PORT);
			_t = new Thread(this);
			_t.setDaemon(true);
			_t.start();
		}catch(IOException e)
		{
			System.out.println("RealmServer : "+e.getMessage());
			Main.agregaralogdemulti("RealmServer : "+e.getMessage());
			Main.agregaralogdeerrores("RealmServer : "+e.getMessage());
			Main.cerrarservidores();
		}
	}

	public void run()
	{
		while(Main.isRunning)
		{
			try
			{
				_clients.add(new RealmThread(_SS.accept()));
			}catch(IOException e)
			{
				try {
					if(!_SS.isClosed()) _SS.close();
					} catch (IOException e1) {}
				System.out.println("RealmServerRun : "+e.getMessage());
				Main.agregaralogdemulti("RealmServerRun : "+e.getMessage());
				Main.agregaralogdeerrores("RealmServerRun : "+e.getMessage());
			}
		}
	}
	
	public void kickAll()
	{
		try
		{
			_SS.close();
		}catch(Exception e)
		{
			System.out.println("RealmServerKickAll : "+e.getMessage());
			Main.agregaralogdemulti("RealmServerKickAll : "+e.getMessage());
			Main.agregaralogdeerrores("RealmServerKickAll : "+e.getMessage());
		}
		ArrayList<RealmThread> c = new ArrayList<>(_clients);
		 for(RealmThread RT : c)
		 {
		 	try
		 	{
		 		RT.closeSocket();
		 	}catch(Exception e){};
	 	}
	}
	
	public void delClient(RealmThread RT)
	{
		_clients.remove(RT);
	}

	public Thread getThread()
	{
		return _t;
	}
}

File: src/main/kotlin/realm/RealmServer.kt
package realm

import common.Config
import java.io.IOException
import java.net.ServerSocket

class RealmServer : Runnable {

    private var _SS: ServerSocket? = null
    private var thread: Thread? = null
    private val _clients = ArrayList<RealmThread>()

    init {
        try {
            _SS = ServerSocket(Config.REALM_PORT)
            thread = Thread(this)
            thread?.isDaemon = true
            thread?.start()
        } catch (e: IOException) {
            println("RealmServer : " + e.message)
            Config.agregaralogdemulti("RealmServer : " + e.message)
            Config.agregaralogdeerrores("RealmServer : " + e.message)
            Config.cerrarservidores()
        }
    }

    override fun run() {
        while (Config.isRunning) {
            try {
                _clients.add(RealmThread(_SS!!.accept()))
            } catch (e: IOException) {
                try {
                    if (!_SS!!.isClosed) _SS!!.close()
                } catch (e1: IOException) {
                }
                println("RealmServerRun : " + e.message)
                Config.agregaralogdemulti("RealmServerRun : " + e.message)
                Config.agregaralogdeerrores("RealmServerRun : " + e.message)
            }
        }
    }

    fun kickAll() {
        try {
            _SS!!.close()
        } catch (e: Exception) {
            println("RealmServerKickAll : " + e.message)
            Config.agregaralogdemulti("RealmServerKickAll : " + e.message)
            Config.agregaralogdeerrores("RealmServerKickAll : " + e.message)
        }
        val c = ArrayList(_clients)
        for (RT in c) {
            try {
                RT.closeSocket()
            } catch (e: Exception) {
            }
        }
    }

    fun delClient(RT: RealmThread) {
        _clients.remove(RT)
    }
}